eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/RulesTests.hs

```diff
@@ -520,3 +542,48 @@ testRwdAliceSignsAlone''' =
         )
         (Coin 0)
         [asWitness Cast.alicePay, asWitness Cast.bobPay]
+
+-- | The reward aggregation bug described in the Shelley ledger spec in
+-- section 17.4 (in the Errata) resulted in needing to use 'aggregatedRewards' to change
+-- the behavior of how rewards are collected starting at protocol version 3.
+-- Instead of collecting a `Coin` for each stake credential, we collect 'Set Reward'.
+-- In major protocol version 2, it is impossible for this set to be empty, but sadly this
+-- property is not enforced in the types. For this reason, the property test
+-- 'propTickfPerservesLedgerView' removes these empty sets from an otherwise arbitrary
+-- 'NewEpochState'.
+filterEmptyRewards :: NewEpochState Shelley -> NewEpochState Shelley
+filterEmptyRewards (NewEpochState el bprev bcur es ru pd stash) =
+  NewEpochState el bprev bcur es ru' pd stash
+  where
+    removeEmptyRewards = Map.filter $ not . Set.null
+    ru' = case ru of
+      SNothing -> SNothing
+      SJust (Pulsing _ _) -> SNothing
+      SJust (Complete rewardUpdate) ->
+        SJust . Complete $ rewardUpdate {rs = removeEmptyRewards (rs rewardUpdate)}
+
+setDepositsToObligation :: NewEpochState Shelley -> NewEpochState Shelley
+setDepositsToObligation nes = nes { nesEs = es { esLState = ls { lsUTxOState =  utxoState }} }
+  where
+    es = nesEs nes
+    ls = esLState es
+    utxoState = (lsUTxOState ls) { utxosDeposited = obligationDPState $ lsDPState ls }
+
+-- | This property test checks the correctness of the TICKF transation.
+-- TICKF is used by the consensus layer to get a ledger view in a computationally
+-- cheaper way than using the TICK rule.
+-- Therefore TICKF and TICK need to compute the same ledger view.
+propTickfPerservesLedgerView :: NewEpochState Shelley -> Property
+propTickfPerservesLedgerView nes =
+  let (EpochNo e) = nesEL nes
+      slot = slotFromEpoch (EpochNo $ e + 1)
+      nes' = setDepositsToObligation (filterEmptyRewards nes)
+      tickNes = runShelleyBase $ applySTSTest @(ShelleyTICK Shelley) (TRC ((), nes', slot))
+      tickFNes = runShelleyBase $ applySTSTest @(ShelleyTICKF Shelley) (TRC ((), nes', slot))
+   in either discard id $ do
+          tickNes' <- tickNes
+          tickNes' <- tickFNes
```

Comment:
@lehins I'll need to undo the suggestion, the return types unfortunately do not match. the trouble is that the predicate failures are different, so the `Left`s are not the same type. I'll revert it to what is was before, unless you have another clever way to make it more readable.